File: src/calculator/n_otc85/cva/trans_pro_calculator.py
from src.const_initializer import ConstInitializer
import numpy as np
import h5py

class TransProCalculator(ConstInitializer):
    """
    Compute trans pro(only content data for visualizer), save relative data

    Methods
    ----------
    save_visual_trans_pro()
        calculator and save relative data

    Attributes
    ---------
    """
    def __init__(self):
        super().__init__()
        # data input
        with h5py.File(self.n_otc85_cva_data_file, 'r') as f:
            self.trans_amp_arr = f[self.trans_amp_name][:]  # general trans pro
        self.trans_pro_arr = np.zeros(
            (self.p_count, self.theta_count, self.visual_count),
            dtype=np.float64,
        )
        self.trans_pro_arr[:, :, 0] = np.abs(self.trans_amp_arr[:, :, 1]) ** 2
        self.trans_pro_arr[:, :, 1] = np.abs(self.trans_amp_arr[:, :, 2] - self.trans_amp_arr[:, :, 1] + self.trans_amp_arr[:, :, 0]) ** 2
        self.trans_pro_arr[:, :, 2] = np.abs(self.trans_amp_arr[:, :, 3]) ** 2
        self.trans_pro_arr[:, :, -1] = np.abs(self.trans_amp_arr[:, :, -1]) ** 2


    def save_visual_trans_pro(self,):
        with h5py.File(self.n_otc85_cva_data_file, 'a') as f:
            if self.trans_pro_name in f:
                del f[self.trans_pro_name]
            trans_pro_dset = f.create_dataset(  # save trans_pro_general
                self.trans_pro_name,
                (self.p_count, self.theta_count, self.visual_count),
                dtype=np.float64,
            )
            trans_pro_dset[:] = self.trans_pro_arr

File: src/calculator/sc/spa/trans_pro_calculator.jl
module TransProCalculator
    include("../../../const_initializer.jl") 
    import .ConstInitializer as CI
    using JLD2

    function save_visual_trans_pro()
        # data input 
        transamp_arr = JLD2.jldopen(CI.sc_spa_datafile, "r") do file
            my_array = read(file, "trans_amp")
            return my_array
        end
        transamp_pro = zeros(Float64, CI.visual_length, CI.plength, CI.thetalength)
        transamp_pro[1, :, :] = @. abs(transamp_arr[2, :, :])^2
        transamp_pro[2, :, :] = @. abs(transamp_arr[3, :, :] - transamp_arr[2, :, :] + transamp_arr[1, :, :])^2
        transamp_pro[3, :, :] = @. abs(transamp_arr[4, :, :])^2
        transamp_pro[4, :, :] = @. abs(transamp_arr[5, :, :])^2
        # add save data
        data = load(CI.sc_spa_datafile)
        data["trans_pro"] = transamp_pro
        save(CI.sc_spa_datafile, data)
    end
end